### ISSUE TITLE
Revise ring buffer code

### ISSUE THREAD

**Issue by twood02** [gwAdvNet2015/adv-net-samples#35]
@wenhuizhang @huyang1988 

I just looked through both of your code more carefully (#30 and #33).  You each have some good parts and some bad parts, so I'd like you to figure out how to fix it up.  Since I've already merged Wenhui's code, it will be best if you can base your changes off of that---in whatever repository you use you should be sure to pull from my repo to be sure you have the latest version of everything and then make your changes to that.  If you followed the instructions in the slides to setup your repo, you should be able to run `git pull upstream master` to pull from my repo into yours.  We set up two remotes: origin=your fork, upstream=the root one in gwAdvNet2015.  You may find it easier to roll your repo back to where it was first forked and then pull (or create a branch from that point).

Here are the issues I see:
- [x] It is good that Wenhui's code uses `void*` as the data type for `items` in `ring.h`. The idea is that the ring buffer should hold a bunch of pointers to arbitrary data. That is why we don't want to just use `int *`. However, at least on my compiler (regular `gcc`), when I try to build the code I get an error, apparently because it doesn't like `(void*) *items`.  It seems that removing the parenthesis fixes the build error and correctly setups a "pointer to a set of pointers".
- [x] There is a difference between how each of you allocate space for the `items` or `buffer` array--compare how you each do it and decide which is correct!
- [x] I hadn't noticed before that Wenhui's test code was using c plus plus.  We don't want that--we want all simple C code for this class.  Yang's code looks closer to what we want. However, you should extend your test cases a bit to be sure everything works--right now you add some items and then remove some items.  Do you know if it will work if you try to add some more after removing?  Also, you should either have comments or better yet print statements in your program that output the expected behavior so someone can easily run the test case and see that it works.  Right now if you run the code you will see things like `push 17 failed`, which will make me think that your code doesn't work.  In fact, it should fail for that since the ring is full.  The program should make that clear.
- [x] Your Makefiles don't actually build your code! I should never have to manually compile anything.  You should have two targets in your file--`ring` and `ringtest`.  The `ringtest` case should depend on `ring`.  Also you should be using gcc since we don't want c++ code. 

See if you can make these changes and submit a new pull request (and get it approved) by thursday night!

Since I approved Wenhui's code the first time around, I'd like Yang to organize the new pull request for this so he also gets practice (but you can work together on the code).

**Comment by wenhuizhang** [comment]
Good evening Prof.  Wood: 

Thank you so much for your comments, the (void*) *items does not work for my gcc compiler as well, and that is why I tried to define a type ptr_t, and it works for my gcc afterwards. 

After changing to void*, I used Visual Studio as IDE, as it is much more convenient to change the format and replace ptr_t, sorry for the fact that I did not notice that the file generated and added to the project folder was a cpp. I was supposed  to delete the last several lines on close the promp and change the file name extension, yet forgot. 

Thank you so much for your review. 

I will get it updated tomorrow when have access to Ethernet. 

Thanks,

Wenhui
